### ISSUE TITLE
Recovery Date overwritten by subsequent checks

### ISSUE THREAD

**Issue by netniV** [Cacti/cacti#1473]
I think the following code is wrong at line 949 in functions.php:
```php
/* if it's time to issue a recovery message, indicate so */
if ($hosts[$host_id]['status_event_count'] >= $ping_recovery_count) {
        /* host is up, flag it that way */
        $hosts[$host_id]['status'] = HOST_UP;

        $issue_log_message = true;

        /* update the recovery date only if the recovery count is 1 */
        if ($ping_recovery_count == $ping_recovery_count) {
                $hosts[$host_id]['status_rec_date'] = date('Y-m-d H:i:s');
        }

        /* reset the event counter */
        $hosts[$host_id]['status_event_count'] = 0;
/* host is recovering, but not ready to issue log message */
} else {
        /* host recovering for the first time, set event date */
        if ($hosts[$host_id]['status_event_count'] == $ping_recovery_count) {
                $hosts[$host_id]['status_rec_date'] = date('Y-m-d H:i:s');
        }
}
```

The comment within the first part of the if clause suggests that we only update the date when the first recovery status is detected.  However, the inner if clause is always true because the variable is checking itself.

I think the code should really be:
```php
/* update the recovery date only if the recovery count is 1 */
if ($hosts[$host_id]['status_event_count'] == 1) {
        $hosts[$host_id]['status_rec_date'] = date('Y-m-d H:i:s');
}

/* if it's time to issue a recovery message, indicate so */
if ($hosts[$host_id]['status_event_count'] >= $ping_recovery_count) {
        /* host is up, flag it that way */
        $hosts[$host_id]['status'] = HOST_UP;

        $issue_log_message = true;

        /* reset the event counter */
        $hosts[$host_id]['status_event_count'] = 0;
}
```

**Comment by pautiina** [comment]
Such a discussion has already been https://github.com/Cacti/cacti/issues/673

**Comment by netniV** [comment]
Yes, but I propose that was still wrong 👍 

If you look, I move the setting of the recovery status to before any check of reovery count.

**Comment by pautiina** [comment]
I looked at the code and agreed with you that the function is not optimal.
But I don't like this line:
`if ($hosts[$host_id]['status_event_count'] == 1) {`
On this line, we ignore the parameter **$ping_recovery_count**

**Comment by netniV** [comment]
Yes. Because the first time the host is in recovery, we set the status date.  Thereafter, the status date is always the date recovery started.

**Comment by pautiina** [comment]
Perhaps this will be more correct:
```

if ($hosts[$host_id]['status_event_count'] == $ping_recovery_count) {
        $hosts[$host_id]['status_rec_date'] = date('Y-m-d H:i:s');
}

/* if it's time to issue a recovery message, indicate so */
if ($hosts[$host_id]['status_event_count'] > $ping_recovery_count) {
        /* host is up, flag it that way */
        $hosts[$host_id]['status'] = HOST_UP;

        $issue_log_message = true;

        /* reset the event counter */
        $hosts[$host_id]['status_event_count'] = 0;
}
```

**Comment by netniV** [comment]
I disagree.  The recovery date should be the date that recovery was started, not when it was notified.

**Comment by pautiina** [comment]
But with this code, the recovery time may not be update in the database, if for some reason spine does not update the field.

**Comment by pautiina** [comment]
We need someone else's opinion. Otherwise, we can make trouble.

**Comment by netniV** [comment]
Take a look at THold's includes/polling.php as that actually uses the status_rec_date field.  It is the time  that the status was achieved.  If you keep updating that time, it gives a false reading that the status was only achieved more recent than it actually was.

So, if you have failure counts set to say 60 with 1 minute polling, the following table is what I believe should happen vs what does happen:

poll # | DateTime | Current | Mine
--- | --- | --- | ---
0 | 16:57:58 | Not Set | Not Set
1 | 16:58:58 | Not Set | 16:58:58
2 | 16:59:58 | Not Set | 16:58:58
3 | 17:00:58 | Not Set | 16:58:58
4 | 17:01:58 | Not Set | 16:58:58
5 | 17:02:58 | Not Set | 16:58:58
6 | 17:03:58 | Not Set | 16:58:58
7 | 17:04:58 | Not Set | 16:58:58
8 | 17:05:58 | Not Set | 16:58:58
9 | 17:06:58 | Not Set | 16:58:58
10 | 17:07:58 | Not Set | 16:58:58
11 | 17:08:58 | Not Set | 16:58:58
12 | 17:09:58 | Not Set | 16:58:58
13 | 17:10:58 | Not Set | 16:58:58
14 | 17:11:58 | Not Set | 16:58:58
15 | 17:12:58 | Not Set | 16:58:58
16 | 17:13:58 | Not Set | 16:58:58
17 | 17:14:58 | Not Set | 16:58:58
18 | 17:15:58 | Not Set | 16:58:58
19 | 17:16:58 | Not Set | 16:58:58
20 | 17:17:58 | Not Set | 16:58:58
21 | 17:18:58 | Not Set | 16:58:58
22 | 17:19:58 | Not Set | 16:58:58
23 | 17:20:58 | Not Set | 16:58:58
24 | 17:21:58 | Not Set | 16:58:58
25 | 17:22:58 | Not Set | 16:58:58
26 | 17:23:58 | Not Set | 16:58:58
27 | 17:24:58 | Not Set | 16:58:58
28 | 17:25:58 | Not Set | 16:58:58
29 | 17:26:58 | Not Set | 16:58:58
30 | 17:27:58 | Not Set | 16:58:58
31 | 17:28:58 | Not Set | 16:58:58
32 | 17:29:58 | Not Set | 16:58:58
33 | 17:30:58 | Not Set | 16:58:58
34 | 17:31:58 | Not Set | 16:58:58
35 | 17:32:58 | Not Set | 16:58:58
36 | 17:33:58 | Not Set | 16:58:58
37 | 17:34:58 | Not Set | 16:58:58
38 | 17:35:58 | Not Set | 16:58:58
39 | 17:36:58 | Not Set | 16:58:58
40 | 17:37:58 | Not Set | 16:58:58
41 | 17:38:58 | Not Set | 16:58:58
42 | 17:39:58 | Not Set | 16:58:58
43 | 17:40:58 | Not Set | 16:58:58
44 | 17:41:58 | Not Set | 16:58:58
45 | 17:42:58 | Not Set | 16:58:58
46 | 17:43:58 | Not Set | 16:58:58
47 | 17:44:58 | Not Set | 16:58:58
48 | 17:45:58 | Not Set | 16:58:58
49 | 17:46:58 | Not Set | 16:58:58
50 | 17:47:58 | Not Set | 16:58:58
51 | 17:48:58 | Not Set | 16:58:58
52 | 17:49:58 | Not Set | 16:58:58
53 | 17:50:58 | Not Set | 16:58:58
54 | 17:51:58 | Not Set | 16:58:58
55 | 17:52:58 | Not Set | 16:58:58
56 | 17:53:58 | Not Set | 16:58:58
57 | 17:54:58 | Not Set | 16:58:58
58 | 17:55:58 | Not Set | 16:58:58
59 | 17:56:58 | Not Set | 16:58:58
60 | 17:57:58 | 17:57:58 | 16:58:58
61 | 17:58:58 | 17:57:58 | 16:58:58
62 | 17:59:58 | 17:57:58 | 16:58:58
63 | 18:00:58 | 17:57:58 | 16:58:58
64 | 18:01:58 | 17:57:58 | 16:58:58
65 | 18:02:58 | 17:57:58 | 16:58:58
66 | 18:03:58 | 17:57:58 | 16:58:58
67 | 18:04:58 | 17:57:58 | 16:58:58
68 | 18:05:58 | 17:57:58 | 16:58:58
69 | 18:06:58 | 17:57:58 | 16:58:58
70 | 18:07:58 | 17:57:58 | 16:58:58

As you can see, that's an hour later than it actually recovered.  If you are using 5 minute polling, that's a lot longer!  @cigamit would weigh in but I'm pretty sure this should be the datetime recovery started, not recovery was emailed otherwise it gives a false impression of outage.

**Comment by pautiina** [comment]
I agree with this:

> As you can see, that's an hour later than it actually recovered. If you are using 5 minute polling, that's a lot longer!

I do not like that if the **status_event_count** is 1, then we do not know that the device was in a state of recovery.

I propose to do so:

```
/* update the recovery date only if the recovery count is 1 */
if ($hosts[$host_id]['status_event_count'] == 1) {
        $hosts[$host_id]['status_rec_date'] = date('Y-m-d H:i:s');
}

/* if it's time to issue a recovery message, indicate so */
if ($hosts[$host_id]['status_event_count'] >= $ping_recovery_count AND $hosts[$host_id]['status_event_count']  != 1) {
        /* host is up, flag it that way */
        $hosts[$host_id]['status'] = HOST_UP;

        $issue_log_message = true;

        /* reset the event counter */
        $hosts[$host_id]['status_event_count'] = 0;
}
```

**Comment by netniV** [comment]
But you do.  By entering that section the host is in a state of recovery as per line 879 because the host status is now being set to UP.  It will only be UP if it's available again, thus recovering.  It's not fully recovered, and full recovery is when the email is sent as the event count is then the host status is changed from RECOVERING to UP.

The following table explains this a bit better based of 5 polling count for recovery:

var status | host status | new status | event count
--- | --- | --- | ---
DOWN | UP | DOWN | 1
UP | DOWN | RECOVERING | 1
UP | RECOVERING | RECOVERING | 2
DOWN | RECOVERING | DOWN | 1
DOWN | DOWN | DOWN | 2
UP | DOWN | RECOVERING | 1
UP | RECOVERING | RECOVERING | 2
UP | RECOVERING | RECOVERING | 3
UP | RECOVERING | RECOVERING | 4
UP | RECOVERING | UP | 0

The recovery date should ALWAYS be set at sequence 1, just as the fail date is, since that is the start of recovery.  The email would only be sent at that final status of UP but with the date of recovery from sequence 1.

**Comment by pautiina** [comment]
Nice table. 
I agree with the logic of the code proposed by you

**Comment by cigamit** [comment]
Down should not occur on first down ping.  There is a setting for that, and the recovery count too.  There may be a bug, this is very old code, and a few people have touched it since 2004, so it might be broken.  I'll explain things, as it is supposed to work, but not tonight.

**Comment by netniV** [comment]
TBH, I wasn't documenting the down process because it was the Recovering/UP part but we could review the entire section.

**Comment by cigamit** [comment]
So, the way it should work is that the device is not formally up until the Recovery Count is reached, and it's not fully down until the Failure Count is reached.  By default they are 3 and 2.  So, you have to wait three polls before the device is actually considered fully up, and 2 for it to be marked down.  When we trigger the event should be at those triggers, when it's marked up or down is another thing altogether.  Likely the real dates as it's better for forensics.

**Comment by pautiina** [comment]
@cigamit I did not fully understand. You are encouraged to use the logic of the proposed @netniV or leave it as is.

**Comment by netniV** [comment]
I read that as he was agreeing with my logic over using the correct recovery date.  However, we do need to make a logic chart to prove what we expect and what is in the code.

**Comment by cigamit** [comment]
@netniV is correct.  I want to use the same approach we used for testing the new graph permissions.  We need to explore all the cases so that we can develop valid test criteria.  In code it's good, be with a state diagram, something I learned in experimental algebra years ago, we can discuss the impact of the design changes as they ripple through the system.

@netniV almost had it right, it needs just a little more detail, like adding the timestamp columns and which is updated by each activity and the up and down intervals (using the defaults), and then we need at least three people in a virtual room to discuss it.  Timing is everything.  I hope to find the time in the next two weeks.
